File: tasks/ppc/slons/solve/solve.py
import requests

import base64
import json
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5


from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.asymmetric import rsa, padding

def restore_message(json_data):
    json_data = base64.b64decode(json_data)
    json_data = json.loads(json_data)
    print(json_data)
    private_key_data = json_data["key"]
    encrypted_message = json_data["task"]["c"]

    private_key_data = private_key_data.encode('utf-8')
    private_key = serialization.load_pem_private_key(private_key_data, password=None, backend=default_backend())

    # Декодируем и расшифровываем сообщение
    encrypted_message = base64.b64decode(encrypted_message)

    # Длина ключа в битах
    key_length = private_key.key_size

    # Разбиваем зашифрованное сообщение на блоки
    num_blocks = len(encrypted_message) // (key_length // 8)

    decrypted_message = b''

    # Расшифровываем каждый блок
    for i in range(num_blocks):
        start = i * (key_length // 8)
        end = (i + 1) * (key_length // 8)
        block = encrypted_message[start:end]
        decrypted_block = private_key.decrypt(block, padding.PKCS1v15())
        decrypted_message += decrypted_block
        print(decrypted_message)
    return decrypted_message.decode('utf-8')

def main():
    from sys import argv
    if(len(argv) < 2):
        print("Usage: python3 solve.py https://example.com")
        return
    url = argv[1]
    session = requests.Session()
    last = session.get(f'{url}/generate-task')
    for i in range(1337):
        ans = restore_message(session.cookies.get("session"))
        print(ans)
        r = session.post(f'{url}/check-task',json={"input": 1},cookies=last.cookies,proxies={'http':"http://localhost:8080"})
        s = r.json()
        if(s['flag']):
            print(s['flag'])
            return

if __name__ == "__main__":
    main()

File: tasks/ppc/slons/solve/solve.js
const NodeRSA = require('node-rsa');
const axios = require('axios');

function restoreMessage(jsonData) {
    const privateKeyData = jsonData.key;
    const encryptedMessage = jsonData.task.c;

    const key = new NodeRSA(privateKeyData, 'pkcs8-private');

    const encryptedBuffer = Buffer.from(encryptedMessage, 'base64');
    const decryptedMessage = key.decrypt(encryptedBuffer, 'utf8');

    return decryptedMessage.split(" ")[2];
}

async function main() {
    const url = process.argv[2];
    if (!url) {
        console.log("Usage: node solve.js https://example.com");
        return;
    }

    const session = axios.create();

    try {
        let lastResponse = await session.get(`${url}/generate-task`);
        for (let i = 0; i < 1337; i++) {
            const ans = restoreMessage(JSON.parse(Buffer.from(lastResponse.headers['set-cookie'][0].split(';')[0].split('=')[1], 'base64').toString()));
            const r = await session.post(`${url}/check-task`, {input: ans}, {
                headers: {
                    'Cookie': lastResponse.headers['set-cookie'],
                    'Content-Type': 'application/json'
                },
                proxy: { host: '127.0.0.1', port: 8080 }
            });

            const s = r.data;
            if (s.flag) {
                console.log(s.flag);
                return;
            }
            lastResponse = r;
        }
    } catch (error) {
        console.error(error);
    }
}

main();
